### DISCUSSION TITLE
IntelliJ IDEA plugin for JBang

### DISCUSSION THREAD

**Discussion by linux-china** [jbangdev/jbang#1135] [Ideas]
I have created IntelliJ IDEA plugin for JBang, and now with following features:

* JDK list sync between IntelliJ IDEA and JBang: sync JDKs from JBang to IntelliJ IDEA
* JSON Schema for jbang-catalog.json:  code completion to edit jbang-catalog.json

More requirements are welcome, then I can release 0.1.0 version. 

![Snip20211218_105](https://user-images.githubusercontent.com/46711/146659920-f693042f-f931-4790-ac3f-bb8558b78a1c.png)

![Snip20211218_106](https://user-images.githubusercontent.com/46711/146659926-b417d642-5874-45b0-85c2-d1b7cc7ea477.png)

**Comment by maxandersen** [comment]
Nice :) let me know if you want me to create a jbangdev/IntelliJ repo and make you commiter :) Upto you. 

How about run/debug via jbang when you detect the ///... on first line ? 

IntelliJ has support for shebang but I don't think it recognizes /// syntax. 

Also content assist for //DEPS lines into maven dependencies would be nice. Similar for other // commands if possible.

**Reply by linux-china** [reply]
@maxandersen  it's good to host plugin on jbangdev/IntelliJ, and it will be nice to create the repo and assign me as commiter.  

I think We don't worry about code completion for //DEPS, and it has been supported by Github Copilot already,  and you can use it with VS Code, Neovim.  Screenshots for IntelliJ IDEA  and Neovim:   

![image](https://user-images.githubusercontent.com/46711/146666636-39daf231-4ce1-4cc0-a1db-36927328ffe7.png)

![image](https://user-images.githubusercontent.com/46711/146666814-e537cc48-c676-43c1-9b8d-2c6fc971d91f.png)

I will add script run feature for the plugin, and it's nice feature.

**Reply by maxandersen** [reply]
I've created https://github.com/jbangdev/jbang-idea and you have an invite to join the github team with commit rights for it

**Reply by maxandersen** [reply]
p.s. I see your current code is kotlin; any objection we make that just .java so easier to contribute/maintain it ?

**Reply by linux-china** [reply]
@maxandersen  the code is based on https://github.com/JetBrains/intellij-platform-plugin-template    Now most JetBrains plugins are written by Kotlin,  and plugin developers can find Kotlin examples from current Kotlin code base easily. Kolin is better for plugin developers, and Rust plugin is written by Kotlin totally.  https://github.com/intellij-rust/intellij-rust/tree/master/src/main/kotlin

**Reply by quintesse** [reply]
Indeed, the IntelliJ eco-system uses a lot of Kotlin and is only going to become more so (they are the creators of Kotlin after all)

**Reply by quintesse** [reply]
Btw @linux-china , I'm not sure I understood your reason for saying code completion on DEPS lines is not needed, can you explain a bit more? I'm not sure what VS Code has to do with code completion in IntelliJ and GitHub codepilot it of course not used by everybody...

**Reply by linux-china** [reply]
yes, I will implement the code completion for DEPS, and it's normal requirement.  I will do some research about code completion in Java comment.

**Comment by linux-china** [comment]
Run configuration support for JBang: run jbang script from IDE

![image](https://user-images.githubusercontent.com/46711/146696790-3637eaf3-d93c-4810-9d01-31248d646069.png)

![Screen Shot 2021-12-19 at 4 36 44 PM](https://user-images.githubusercontent.com/46711/146696829-a5e68ebb-db8e-4f90-b22b-433bbca9433f.png)

**Reply by quintesse** [reply]
Nice!

Should we perhaps also add a "New -> Jbang script" option? (I always forget what the exact first line is so it would be useful for me)

Could you perhaps add a small explanation to the [README](https://github.com/jbangdev/jbang-idea/blob/main/README.md) on how to install the plugin?

**Reply by linux-china** [reply]
Yes, it's a must. New script file will include jbang declaration and be executable automatically on Mac and Linux. 

<img width="370" alt="Screen Shot 2021-12-20 at 1 46 43 PM" src="https://user-images.githubusercontent.com/46711/146837198-b6de74cb-fb00-4bf3-b267-b43ff4ee264f.png">

<img width="566" alt="Screen Shot 2021-12-20 at 12 41 41 PM" src="https://user-images.githubusercontent.com/46711/146837223-85e3329c-c837-4f02-8556-93c799f5671f.png">

**Reply by quintesse** [reply]
Awesome! :smile:  :+1:

**Reply by quintesse** [reply]
Btw I see you wrote:

```
Install
Preferences > Plugins > Marketplace > Search for "jbang" > Install Plugin > Restart IntelliJ IDEA
```

In the README. Should that work already? Or should we publish that plugin first? (Because searching for "jbang" doesn't give any results for me right now)

**Reply by linux-china** [reply]
I will release 0.1.0 version ASAP. Now you can try it by install from Disk. 

The following features are not included in 0.1.0 version because of some hard.

* GAV completion for //DEPS
* Classpath setting for single Java file from //DEPS


[jbang-intellij-plugin-0.1.0.zip](https://github.com/jbangdev/jbang/files/7748323/jbang-intellij-plugin-0.1.0.zip)

**Reply by linux-china** [reply]
@maxandersen  @quintesse   To be honest,  English is not my first language,  please correct README.md if spell error.

**Reply by maxandersen** [reply]
Where is it actually published/hosted from ?

**Reply by linux-china** [reply]
I have submitted the plugin to https://plugins.jetbrains.com/      It will be available to public after approved, and almost 2 days.

**Comment by linux-china** [comment]
Directive completion ready now, and I will do more research to implement GAV completion for //DEPS

<img width="458" alt="Screen Shot 2021-12-21 at 2 48 46 PM" src="https://user-images.githubusercontent.com/46711/147007866-67f4cbf3-9754-42c4-bb20-a665ae652cfa.png">

<img width="458" alt="Screen Shot 2021-12-21 at 2 48 11 PM" src="https://user-images.githubusercontent.com/46711/147007881-720e7c68-f7e1-4b5f-beeb-4d0c41ec7cfb.png">

**Reply by maxandersen** [reply]
how do you get these completions btw? I don't see them...

**Comment by linux-china** [comment]
I have uploaded the plugin to plugins.jetbrains.com.   Xmas is coming,  and I think JetBrains guys will have long holiday.  Expect to get approved before Xmas.  If some important bugs to the plugin, and I will release a new version ASAP. 

![image](https://user-images.githubusercontent.com/46711/147013306-b401824b-3ee7-47f3-8d97-b102d073e5d1.png)


The plugin url:  https://plugins.jetbrains.com/plugin/18257-jbang

**Reply by maxandersen** [reply]
Cool. Is that using a system where we can get a jbangdev org/user to have the credentials to push so we can automate it from GitHub actions ?

**Reply by linux-china** [reply]
Yes,  we can automate it from GitHub actions.   You should upload first version by hand, then you can update the plugin automatically.  I will setup it later, and it's my first time to try automate too.  :)  

```
  publishPlugin {
        dependsOn("patchChangelog")
        token.set(System.getenv("PUBLISH_TOKEN"))
        // pluginVersion is based on the SemVer (https://semver.org) and supports pre-release labels, like 2.1.7-alpha.3
        // Specify pre-release label to publish the plugin in a custom Release Channel automatically. Read more:
        // https://plugins.jetbrains.com/docs/intellij/deployment.html#specifying-a-release-channel
        channels.set(listOf(properties("pluginVersion").split('-').getOrElse(1) { "default" }.split('.').first()))
    }
```

**Reply by maxandersen** [reply]
cool - I got a little bit of help and its now public at https://plugins.jetbrains.com/plugin/18257-jbang
but for some reason I can't install it into latest community edition ...which version of intellij does it need ?

**Reply by nandorholozsnyak** [reply]
Just downloaded it! Nice work! Will provide feedback as soon as I can! :)

**Reply by manusa** [reply]
I haven't tested it yet, but from what I've read in the other discussions + release notes, I'm unsure if it configures the dependencies for you (+ recalculate on //DEPS change/add/remove).

This is basically the feature I miss most. My usual workflow involves opening/editing as Gradle project, changing dependencies, closing, opening, GOTO 10, ...

**Reply by manusa** [reply]
BTW, congrats on the release and great job!

**Reply by linux-china** [reply]
@manusa   I created the issue:  sync dependencies between build.gradle and JBang script   https://github.com/jbangdev/jbang-idea/issues/3   If not correct, please correct me there.

**Reply by manusa** [reply]
Thx, I think it's OK. I'll also try to test the plugin and provide feedback.

Maybe I'm confused about how we should use it to edit/create scripts.

e.g. `$ idea MyScript.java`

**Reply by maxandersen** [reply]
It's 'jbang edit my script.java' then a temporary project gets setup you can edit with moats IDEs

**Reply by manusa** [reply]
Yes, so this is my usual workflow:
* `jbang edit --open='/path/to/intellij/idea.sh' --live MyScript.java` (choose Gradle Project)
* Modify the `//DEPS ` section
* Save and close project
* Go to first step (Gradle project gets regenerated, thus modified deps are listed as `External dependencies`)

**Reply by manusa** [reply]
:facepalm: OK, my bad, the `--live` covered for what I was asking.

**Reply by maxandersen** [reply]
You can use 'jbang edit --live script.java' then jbang will continuously update the ide files. 

You still in idea need to manually refresh but no need to exit. 

Btw. 0.85 we just released let's you do 'jbang config set edit.open <path to idea>'
So you don't have to do that every time. 

Hope that helps!

**Reply by manusa** [reply]
Thx, Max.
Yes, the `--live` flag covered for what I needed when editing an individual script.
Sorry for the noise.

So I have tested the plugin now and I think I can provide a better "user story" or use-case of sorts.
I think this is applicable exclusively for IDE users _(---> has to be implemented in an IDE plugin)_, as said, the `jbang edit --live` covers the dependency issue for single files.

--- 

As a user who has a directory with **multiple** JBang scripts, or a (Java) project that contains a directory with JBang scripts, I want to be able to work with them without exiting the IDE.

```shell
$ idea /path/to/project/or/script/directory
``` 

- :heavy_check_mark:  Create a new Script
  - _This seems to be already covered `New > JBang` script_
- :heavy_check_mark: Run a script
  - _There is a JBang run configuration available_
  - _`Right click > Run` automatically detects the JBang script_
- :heavy_multiplication_x: Debug a script 
  - _JBang run configurations don't offer a debug option_
- :heavy_multiplication_x: Code completion for script
  - _Need to manually setup base JDK version (once)_
  - _Dependencies are not detected for the individual script_ 

--- 

IMHO improving this scenario would be great. It would enhance my experience when dealing with quickstarts/examples for my project (I think JBang is great for this, but is tedious to maintain them). Same applies for project related bash scripts that could be easily replaced by JBang, etc.

**Reply by maxandersen** [reply]
If you uninstall the jbang plugin debugging is there when you use jbang edit. 

We need to fix it so the plug-in doesn't remove that. 

Jbang edit for directories (which would mean all their dependencies must be added to one gradle file because of IDE limitations today) is in works. 
GitHub mobile client doesn't give me autocomplete so can't link it it but it's in one of the open pull requests.

**Reply by manusa** [reply]
> [...] which would mean all their dependencies must be added to one gradle file because of IDE limitations today [...]

This is why most of my comment belongs and applies to the IntelliJ plugin and not the jbang edit command which is fine as it is (or maybe can be improved with what you mention -- _jbang edit directory_).

Your statement is not exactly true for IntelliJ. I don't know the internals, but having a single dependency set only applies to the global **idea** project/workspace. This is what is displayed as `External Libraries` in your `Project` window (bottom).

However, an **idea** project/workspace can have any number of nested configured Maven and/or Gradle projects. You can add any number of those to your project by using the `Add Maven Projects` or `Link Gradle Project` actions (or their Drag and drop replacements).

For each of the configured Gradle/Maven projects (+ their sub-modules) there's a tree showing the applicable dependencies (on the corresponding Maven/Gradle window). These dependencies are also added to the global **idea** project External Libraries section (this is what happens when you press the `Reload All Maven/Gradle Projects` button.

So for `jbang edit` adding every dependency to a single Gradle project sure is the easiest approach. But I don't think this would be the right approach for the plugin.

IMHO, regarding the plugin, basically having JBang files behave like independent nested projects, i.e. mimicking what Gradle and Maven plugins already do, would be more or less how to achieve the multi-jbang file support for a single **idea** project/workspace.

Ideally, the plugin would provide a window where you would configure the applicable JBang files (just like you do with nested Maven and Gradle projects). For each entry, we can see the JBang script dependencies, which in turn are added to the global idea project/workspace `External Libraries`.

Again, this is on the premise that you want the plugin to work when loading a project/directory using `$ idea /path/to/dir` and not `$ jbang edit`.

So let's say I have my nice Maven project which contains a few JBang utility scripts. I could then configure those in my **idea** project too and would be able to edit/execute them just like any other Java file in the main Maven project.

**Reply by maxandersen** [reply]
Are you suggesting to setup or rather mimick a project per .java file inside IntelliJ ?

**Reply by maxandersen** [reply]
Note: I don't think that will ever work well. IntelliJs internals are reliant on physical folders.

**Reply by maxandersen** [reply]
If it can - great; but I honestly think that will be worse than just having jbang plug-in treat directory open same as what jbang edit . Would do. 

Too many impedance mismatches trying to get UX of the IDE and user to work meaningfully.

**Reply by manusa** [reply]
> Are you suggesting to setup or rather mimick a project per .java file inside IntelliJ ?

Sth like that. Yes.

Anyway, I'm just talking from the UX perspective because I haven't ever maintained/created an IntelliJ plugin.

I hope I can find some time to play with this and maybe provide sth myself.

**Reply by linux-china** [reply]
How about multi modules?  each module maintains Java files and dependencies.  I implemented an action to sync DEPS to module's dependencies, and developers don't worry about dependencies management for module.

![Screen Shot 2021-12-28 at 1 18 34 AM](https://user-images.githubusercontent.com/46711/147550544-b98b9ff2-35b4-47e6-ac03-6fbe51d318fd.png)

**Reply by maxandersen** [reply]
I don't follow what you mean @linux-china ? To have multi modules you need to have multiple build files / folders don't you ?

**Reply by maxandersen** [reply]
I tried latest where you sync jbang deps to a idea module - which works "ok" but how do you expect that to work with multiple files?

**Reply by maxandersen** [reply]
I opened issues on jbang-idea - the sync is very broken IMO. it doesn't seem to sync from build.gradle nor the .java file it simply just injects what is in the gradle dependency container and pushes that into your .java file. Which in majority of  cases is not what you want.

**Reply by linux-china** [reply]
> I don't follow what you mean @linux-china ? To have multi modules you need to have multiple build files / folders don't you ?

Yes, it's hard to fix this problem in IDEA with multi JBang script files.   Every script has different JAVA version and dependencies, and  you should create a IDEA module for the script.   But still ok, IDEA's module is same as SourceSet in Gradle,  different SourceSet has different dependencies and language level.    But don't worry,  IDEA's module and Gradle SourceSet are very light. 

I don't know it's good or not to ask the developers to group scripts together as SourceSet, and yes SourceSet is a folder.

**Reply by linux-china** [reply]
> I opened issues on jbang-idea - the sync is very broken IMO. it doesn't seem to sync from build.gradle nor the .java file it simply just injects what is in the gradle dependency container and pushes that into your .java file. Which in majority of cases is not what you want.

Version 0.6.0 is better and still waiting for approval.

**Reply by maxandersen** [reply]
Im running using latest build 0.7 from main branch

**Reply by linux-china** [reply]
Some test cases here, and if not correct please correct me. 

* Open JBang projedct with `idea .` without build.gradle:   sync dependencies from `//DEPS ` to IDEA's module, just one way
* `jbang edit --open=idea HelloWorld.java`:  sync dependencies between `//DEPS ` and build.gradle ,  translate `implementation` to `//DEPS ` and reverse
*  Gradle project with multi SourceSets:    sync dependencies between `//DEPS ` and build.gradle ,  translate `xxxImplementation` to `//DEPS ` and reverse   

Some issues:

* `implementation platform('bom-here')`   can not work well and I'm working on it
*  Refresh automatically:  now you need click Gradle refresh button
* jbang.jar bundle can not work well because of classloader confliction

I'm working on a tool window for JBang, and display all information for script from `jbang info tools src/HelloWorld.java`  and some info from gradle. If not correct and developers can click sync button again.

**Reply by maxandersen** [reply]
I would just do //DEPS to other Modules and make that work well. The reverse way from build.Gradle is lossy.

**Comment by linux-china** [comment]
High lighter for JBang directives.   If not necessary, and I can disable the feature. 

<img width="771" alt="Screen Shot 2021-12-31 at 5 48 07 PM" src="https://user-images.githubusercontent.com/46711/147842443-fc85716e-cebe-48e9-99da-d92d86d6b9e7.png">

**Reply by maxandersen** [reply]
looks nice!

**Comment by nandorholozsnyak** [comment]
Any news on a new release? I'm able to build it from the code, just curious if there will be new updates in the plugins store as well.
On the other hand I know I should create an issue for that, but I am unable to run a JBang script within IDEA, when I click on the icon or just create a run config and try to run the specific file, the following error happens, this is what I see in the logs of IDEA:
Plugin version: 0.8.0
JBang version: 0.86.0
```
2022-01-08 13:38:53,086 [ 981711]   INFO - xecution.runners.ExecutionUtil - Error running 'RestEasyExample.java by JBang':<br>Cannot run program "jbang" (in directory "/home/nandi/EVCS/VCS/GIT/jbang-terraform-aws"): error=2, No such file or directory 
com.intellij.execution.process.ProcessNotCreatedException: Cannot run program "jbang" (in directory "/home/nandi/EVCS/VCS/GIT/jbang-terraform-aws"): error=2, No such file or directory
	at com.intellij.execution.configurations.GeneralCommandLine.createProcess(GeneralCommandLine.java:364)
	at com.intellij.execution.process.OSProcessHandler.startProcess(OSProcessHandler.java:94)
	at com.intellij.execution.process.OSProcessHandler.<init>(OSProcessHandler.java:46)
	at com.intellij.execution.process.KillableProcessHandler.<init>(KillableProcessHandler.java:41)
	at com.intellij.execution.process.ColoredProcessHandler.<init>(ColoredProcessHandler.java:23)
	at com.intellij.execution.process.ProcessHandlerFactoryImpl.createColoredProcessHandler(ProcessHandlerFactoryImpl.java:33)
	at dev.jbang.idea.run.JBangRunConfiguration$getState$1.startProcess(JBangRunConfiguration.kt:78)
	at com.intellij.execution.configurations.CommandLineState.execute(CommandLineState.java:61)
	at com.intellij.execution.runners.DefaultRunProgramRunner$execute$1.invoke(DefaultRunProgramRunner.kt:29)
	at com.intellij.execution.runners.DefaultRunProgramRunner$execute$1.invoke(DefaultRunProgramRunner.kt:12)
	at com.intellij.execution.impl.ExecutionManagerImpl$startRunProfile$1.invoke(ExecutionManagerImpl.kt:173)
	at com.intellij.execution.impl.ExecutionManagerImpl$startRunProfile$1.invoke(ExecutionManagerImpl.kt:72)
	at com.intellij.execution.impl.ExecutionManagerImpl$doStartRunProfile$startRunnable$1.run(ExecutionManagerImpl.kt:234)
	at com.intellij.openapi.application.TransactionGuardImpl.runWithWritingAllowed(TransactionGuardImpl.java:214)
	at com.intellij.openapi.application.TransactionGuardImpl.access$200(TransactionGuardImpl.java:21)
	at com.intellij.openapi.application.TransactionGuardImpl$2.run(TransactionGuardImpl.java:196)
	at com.intellij.openapi.application.impl.ApplicationImpl.runIntendedWriteActionOnCurrentThread(ApplicationImpl.java:805)
	at com.intellij.openapi.application.impl.ApplicationImpl.lambda$invokeLater$4(ApplicationImpl.java:348)
	at com.intellij.openapi.application.impl.FlushQueue.doRun(FlushQueue.java:82)
	at com.intellij.openapi.application.impl.FlushQueue.runNextEvent(FlushQueue.java:131)
	at com.intellij.openapi.application.impl.FlushQueue.flushNow(FlushQueue.java:47)
	at com.intellij.openapi.application.impl.FlushQueue$FlushNow.run(FlushQueue.java:187)
	at java.desktop/java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:313)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:776)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:727)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:721)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:746)
	at com.intellij.ide.IdeEventQueue.defaultDispatchEvent(IdeEventQueue.java:891)
	at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.java:760)
	at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$6(IdeEventQueue.java:447)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computePrioritized(CoreProgressManager.java:818)
	at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$7(IdeEventQueue.java:446)
	at com.intellij.openapi.application.impl.ApplicationImpl.runIntendedWriteActionOnCurrentThread(ApplicationImpl.java:805)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:492)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)
Caused by: java.io.IOException: Cannot run program "jbang" (in directory "/home/nandi/EVCS/VCS/GIT/jbang-terraform-aws"): error=2, No such file or directory
	at java.base/java.lang.ProcessBuilder.start(ProcessBuilder.java:1128)
	at java.base/java.lang.ProcessBuilder.start(ProcessBuilder.java:1071)
	at com.intellij.execution.configurations.GeneralCommandLine.startProcess(GeneralCommandLine.java:440)
	at com.intellij.execution.configurations.GeneralCommandLine.createProcess(GeneralCommandLine.java:354)
	... 41 more
Caused by: java.io.IOException: error=2, No such file or directory
	at java.base/java.lang.ProcessImpl.forkAndExec(Native Method)
	at java.base/java.lang.ProcessImpl.<init>(ProcessImpl.java:340)
	at java.base/java.lang.ProcessImpl.start(ProcessImpl.java:271)
	at java.base/java.lang.ProcessBuilder.start(ProcessBuilder.java:1107)
	... 44 more

```

**Reply by maxandersen** [reply]
Where is your jbang installed ?

With respect to releases - awaiting update from @linux-china on moving jbang plugin to jbangdev at jetbrains - then we can start automatically publish it

**Reply by nandorholozsnyak** [reply]
> Where is your jbang installed ?
> 
> With respect to releases - awaiting update from @linux-china on moving jbang plugin to jbangdev at jetbrains - then we can start automatically publish it

```
❯ which jbang
/home/nandi/.jbang/bin/jbang
```

**Reply by maxandersen** [reply]
That's weird as the code actually looks exactly there. Can you open a bug on jbang-idea ?

**Reply by linux-china** [reply]
@nandorholozsnyak   It is caused by jbang not in PATH.   If you start the project by idea command, for example `idea .`, and it will work well. 

```
val command = mutableListOf("jbang", "run")
```

I have fixed this problem to use jbang cmd full path, and new version will be released ASAP

**Reply by linux-china** [reply]
Because of long holiday, and version 0.7.0 is still waiting for approval and it's almost 2 weeks.   I just uploaded new version of 0.8.0 

<img width="1400" alt="Screen Shot 2022-01-08 at 1 08 25 PM" src="https://user-images.githubusercontent.com/46711/148660373-489a3c16-e79e-4fbf-b26e-a0245fa13169.png">

**Reply by nandorholozsnyak** [reply]
Thank you. I should have checked my PATH. Will check the new version as soon as I can.

**Reply by nandorholozsnyak** [reply]
BTW, this issue for me is gone on both Windows and Linux (Pop OS) too. Thank you! 
I'm at 0.10.0

**Comment by linux-china** [comment]
GAV completion finished in 0.11.0 version 

![Snip20220118_212](https://user-images.githubusercontent.com/46711/150037458-edd00d4d-ccd6-4a5e-8432-deae35da2c46.png)

**Reply by nandorholozsnyak** [reply]
Awesome!!

**Reply by quintesse** [reply]
Indeed, awesome! 👏

**Reply by nandorholozsnyak** [reply]
Just tried it out, superb! Probably the most convenient feature! Congrats.

**Reply by maxandersen** [reply]
I don't get any auto completion - anything magic to type/press to have it show up ?

**Reply by nandorholozsnyak** [reply]
Did you update to 0.11.0? Of yes then start typing for example quarkus and it should show the possible dependencies. Or this is what I did at least.

**Reply by maxandersen** [reply]
Yes. I'm on 0.11 but nothing shows up :/

**Reply by maxandersen** [reply]
I'll try use it from latest source tomorrow.

**Comment by nandorholozsnyak** [comment]
Just a quick question, is there anybody whos IDEA also freezes on the Sync action? Is there any way to make it an async background task to sync the dependencies?
https://github.com/jbangdev/jbang-idea/issues/19

**Comment by nandorholozsnyak** [comment]
Hey @linux-china  are you planning to enable/add Debug button feature in the plugin?
Right now if I want to debug it I have to give the `--debug` flag to the run config and create a Remote JVM Debug config and run it after the JBang script run.

**Reply by linux-china** [reply]
@nandorholozsnyak  Now you can use Debug supplied by IDE. After dependencies synced to IDEA's module, and you can run JBang scripts as normal Java app.   Now I use this way to debug JBang scripts.   My JBang app structure as following: 

```
jbang-app
  - src
  - resources
  - tests
  - jbang-catalog.json
```

I can implement debug feature, but I don't how many people will like this feature, and maybe most developers like to debug JBang script as normal Java/Groovy/Kotlin script in IDEA.

**Reply by nandorholozsnyak** [reply]
Cool thank you. I mostly use/create Quarkus based apps and in these cases I have to override the main entry point, with the debug it maybe a bit easier, but that is fine. And on the other hand, would the Quarkus DEV mode work if I would use a main method entrypoint? It has to be triggered with the -Dquarkus.dev flag when running a jbang script.

**Reply by maxandersen** [reply]
Debug via jbang would be nice as right now debug via just java does not process //FILES. 

Wonder if jbang could register as something to run before java run/debug kicks in ?
